### ISSUE TITLE
Incorrect inheritance checking in Exception handler code

### ISSUE THREAD

**Issue by bk5115545** [Kludex/starlette#2273]
Given a class created with a metaclass while also inheriting from Exception, https://github.com/encode/starlette/blob/master/starlette/_exception_handler.py#L17 does not properly consider this as an exception and therefore fails the below code:

```python
class MyMeta(type):
    def __new__(cls, name, bases, dct):
        return super().__new__(cls, name, bases, dct)

class MyOutputParent(metaclass=MyMeta):
    ...

class MyExceptionParent(MyOutputParent, Exception):
    ...

print(f'{(Exception in type(MyExceptionParent).__mro__)=}')  # Fails (this is what the exception middleware does)
print(f'{(Exception in MyExceptionParent.__mro__)=}')  # passes
print(f'{isinstance(MyExceptionParent(), Exception)=}')  # passes
print(f'{issubclass(MyExceptionParent, Exception)=}')  # passes
```

This prevents custom serialization frameworks that deal with both exceptions and metaclasses from using the built-in `app.add_exception_handler` effectively.

Found on 0.23 however still present on master

> [!IMPORTANT]
> - We're using [Polar.sh](https://polar.sh/encode) so you can upvote and help fund this issue.
> - We receive the funding once the issue is completed & confirmed by you.
> - Thank you in advance for helping prioritize & fund our backlog.

<a href="https://polar.sh/encode/starlette/issues/2273">
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://polar.sh/api/github/encode/starlette/issues/2273/pledge.svg?darkmode=1">
  <img alt="Fund with Polar" src="https://polar.sh/api/github/encode/starlette/issues/2273/pledge.svg">
</picture>
</a>

**Comment by kpberry** [comment]
I tried to reproduce this issue, but it seems that app.add_exception_handler correctly handles the MyExceptionParent you described. Here is the app I used for testing:

```python
from starlette.applications import Starlette
from starlette.responses import PlainTextResponse
import uvicorn


# copied from bk5115545's comment
class MyMeta(type):
    def __new__(cls, name, bases, dct):
        return super().__new__(cls, name, bases, dct)

class MyOutputParent(metaclass=MyMeta):
    ...

class MyExceptionParent(MyOutputParent, Exception):
    ...
# end copied section


def raise_exception(req):
    raise MyExceptionParent()


def handle_exception(*args, **kwargs):
    return PlainTextResponse('exception handled')


app = Starlette()
app.add_route('/raise_exception', raise_exception, methods=['GET'])
app.add_exception_handler(MyExceptionParent, handle_exception)

uvicorn.run(app)
```

When I submit a GET request to the /raise_exception endpoint, I get the response "exception handled", indicating that the exception handler is running for instances of MyExceptionParent as expected.

**Comment by aminalaee** [comment]
Yeah I think @kpberry  is right.

The assumption in the issue does not seem correct.
Starlette does `type(MyExceptionParent())` not `type(MyExceptionParent)`.
Because:

```py
type(MyExceptionParent) # MyMeta
type(MyExceptionParent()) # MyExceptionParent
```

And:

```py
print(f'{(Exception in type(MyExceptionParent()).__mro__)=}') # passes
```

**Comment by bk5115545** [comment]
I think I should have included the `add_exception_handler` call.

In my current code, adding a handler for the parent itself works fine however adding a handler for `Exception` doesn't catch exceptions with a metaclass.

Does not work for exceptions with a metaclass:
```
app.add_exception_handler(Exception, exception_handler)
```

Works:
```
app.add_exception_handler(MyExceptionParent, exception_handler)
```

**Comment by bk5115545** [comment]
Trying your @kpberry's code directly shows that it is handling correctly using my example.

I am probably not understanding how our close-source framework is creating Exceptions.

Making a slight modification to the code:
```python
from starlette.applications import Starlette
from starlette.responses import PlainTextResponse
import uvicorn

from closed.source import ClosedSourceException # this is roughly MyParentException

def raise_exception(req):
    raise ClosedSourceException(message='test')


def handle_exception(*args, **kwargs):
    return PlainTextResponse('exception handled')


app = Starlette()
app.add_route('/raise_exception', raise_exception, methods=['GET'])
app.add_exception_handler(Exception, handle_exception)

uvicorn.run(app)
```

I get the below output:
```
INFO:     127.0.0.1:56445 - "GET /raise_exception HTTP/1.1" 200 OK
ERROR:    Exception in ASGI application
Traceback (most recent call last):
  File "...python3.9/site-packages/uvicorn/protocols/http/h11_impl.py", line 364, in run_asgi
    result = await app(self.scope, self.receive, self.send)
  File "...python3.9/site-packages/uvicorn/middleware/proxy_headers.py", line 75, in __call__
    return await self.app(scope, receive, send)
  File "...python3.9/site-packages/starlette/applications.py", line 125, in __call__
    await self.middleware_stack(scope, receive, send)
  File "...python3.9/site-packages/starlette/middleware/errors.py", line 184, in __call__
    raise exc
  File "...python3.9/site-packages/starlette/middleware/errors.py", line 162, in __call__
    await self.app(scope, receive, _send)
  File "...python3.9/site-packages/starlette/middleware/exceptions.py", line 79, in __call__
    raise exc
  File "...python3.9/site-packages/starlette/middleware/exceptions.py", line 68, in __call__
    await self.app(scope, receive, sender)
  File "...python3.9/site-packages/starlette/routing.py", line 706, in __call__
    await route.handle(scope, receive, send)
  File "...python3.9/site-packages/starlette/routing.py", line 276, in handle
    await self.app(scope, receive, send)
  File "...python3.9/site-packages/starlette/routing.py", line 68, in app
    response = await run_in_threadpool(func, request)
  File "...python3.9/site-packages/starlette/concurrency.py", line 41, in run_in_threadpool
    return await anyio.to_thread.run_sync(func, *args)
  File "...python3.9/site-packages/anyio/to_thread.py", line 28, in run_sync
    return await get_asynclib().run_sync_in_worker_thread(func, *args, cancellable=cancellable,
  File "...python3.9/site-packages/anyio/_backends/_asyncio.py", line 802, in run_sync_in_worker_thread
    return await future
  File "...python3.9/site-packages/anyio/_backends/_asyncio.py", line 740, in run
    result = func(*args)
  File "<stdin>", line 2, in raise_exception
closed.source.ClosedSourceException <ClosedSourceException(message='test')>
```

```
>>> from closed.source import ClosedSourceException
>>> isinstance(ClosedSourceException(), Exception)
True
>>> issubclass(ClosedSourceException, Exception)
True
>>> type(ClosedSourceException()).__mro__
(<class 'closed.source.ClosedSourceException'>, <class 'closed.source.structure.Structure'>, <class 'closed.source.helpers.FromJsonMixIn'>, <class 'Exception'>, <class 'BaseException'>, <class 'object'>)
```

**Comment by kpberry** [comment]
Adding a handler for Exception (rather than any subclass of Exception) is a special case in Starlette, and overrides the base error handler. Thus, it uses the ServerErrorMiddleware instead of the ExceptionMiddleware. From the [documentation](https://www.starlette.io/exceptions/):

> It is important to differentiate between handled exceptions and errors.
Handled exceptions do not represent error cases. They are coerced into appropriate HTTP responses, which are then sent through the standard middleware stack. By default the HTTPException class is used to manage any handled exceptions.
Errors are any other exception that occurs within the application. These cases should bubble through the entire middleware stack as exceptions. Any error logging middleware should ensure that it re-raises the exception all the way up to the server.
In practical terms, the error handled used is exception_handler[500] or exception_handler[Exception]. Both keys 500 and Exception can be used.

In other words, you are seeing the error logging in the console because the ServerErrorMiddleware first handles the error, sending "exception handled" as a response, then re-raises the error to be processed by the rest of the middleware stack.

If you don't want the error propagation, you can add a handler for a subclass of Exception, such as MyExceptionParent. If you want all instances of your metaclass to be handled Exceptions, you could add a subclass of Exception to the bases list in `MyMeta.__new__` and add a handler for that Exception subclass, though you would need to be careful to not accidentally generate an inconsistent mro when `__new__` is actually called.

**Comment by bk5115545** [comment]
Thank you for the details.

The MyMeta and the exceptions are actually generated via another framework using Smithy specification and implemented as Rust with Py03 bindings. While they are valid Python exceptions, they're not something that is easy to change

Most likely, I will do some `__import__` nonsense to recursively import all exceptions in my module and register them using `add_exception_handler`. I really do not like the ergonomic impact of this solution but I can't see another way
